### PR TITLE
Remove public access modifier for parameters

### LINKED ISSUE #25204: Remove public for parameters
**Description:**
Lang issue to track https://github.com/ballerina-platform/ballerina-spec/issues/577

### PR DESCRIPTION
**Pull request by dulvinw**
Fixes #25204

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by dulvinw** [comment]
Will be adding testcases

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 1]
```diff
@@ -283,10 +283,10 @@
             { ParserRuleContext.CLOSE_PARENTHESIS, ParserRuleContext.REQUIRED_PARAM };
 
     private static final ParserRuleContext[] PARAMETER_START =
-            { ParserRuleContext.ANNOTATIONS, ParserRuleContext.PUBLIC_KEYWORD, ParserRuleContext.TYPE_DESC_IN_PARAM };
+            { ParserRuleContext.ANNOTATIONS, ParserRuleContext.TYPE_DESC_IN_PARAM };
 
     private static final ParserRuleContext[] PARAMETER_WITHOUT_ANNOTS =
-            { ParserRuleContext.PUBLIC_KEYWORD, ParserRuleContext.TYPE_DESC_IN_PARAM };
+            { ParserRuleContext.TYPE_DESC_IN_PARAM };
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:289]: With this change we don't have alternative paths here ryt. Can remove `PARAMETER_WITHOUT_ANNOTS` as an alternative path and give `TYPE_DESC_IN_PARAM` as the next rule directly.

Comment (dulvinw) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:289]: +1 will do that

File: stdlib/database/sql/src/test/resources/test-module/mock-sql/src/mockclient/resources/Client_functions.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"testorg/mockclient:1.0.0:Client" : [{"function":{"name":"query","requiredParams":[{"sqlQuery":"(string|ballerina/sql:0.5.0:ParameterizedQuery)"},{"rowType":"typedesc<record {| anydata...; |}>?"}],"returnType":"stream"}},{"function":{"name":"execute","requiredParams":[{"sqlQuery":"(string|ballerina/sql:0.5.0:ParameterizedQuery)"}],"returnType":"ballerina/sql:0.5.0:ExecutionResult"}},{"function":{"name":"batchExecute","requiredParams":[{"sqlQueries":"ballerina/sql:0.5.0:ParameterizedQuery[]"}],"returnType":"ballerina/sql:0.5.0:ExecutionResult[]"}}]}
```
Comment (MaryamZi) [inline stdlib/database/sql/src/test/resources/test-module/mock-sql/src/mockclient/resources/Client_functions.json:1]: Is this file required?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 3]
```diff
@@ -1466,6 +1467,10 @@ private void checkForExportableType(BTypeSymbol symbol, DiagnosticPos pos) {
                 }
                 checkForExportableType(invokableType.retType.tsymbol, pos);
                 return;
+            case TypeTags.PARAMETERIZED_TYPE:
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:1470]: Was there something specific that failed, requiring this change?

Comment (dulvinw) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:1470]: Yes actually. I was getting an error for trying to export a non public symbol for something like 
```function test(typedesc<object {}> T) returns T {}```

File: stdlib/database/sql/src/test/resources/test-module/mock-sql/src/mockclient/resources/Client_functions.json (hunk 2 above)
Comment (dulvinw) [reply stdlib/database/sql/src/test/resources/test-module/mock-sql/src/mockclient/resources/Client_functions.json:1]: This was autogenerated.

Comment (dulvinw) [reply stdlib/database/sql/src/test/resources/test-module/mock-sql/src/mockclient/resources/Client_functions.json]: Fixed in 785c34a

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 1 above)
Comment (dulvinw) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java]: Fixdd in 785c34a

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 3 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:1470]: Had an offline discussion with @dulvinw regarding this too, previously `typedesc<object {}> T` in this same example was marked as a public param, which is why we didn't get the error.

We will go ahead with this change.

**Review by lochana-chathura — CHANGES_REQUESTED** [review]
BallerinaParserErrorHandler changes are OK!

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -1302,7 +1302,6 @@ private STNode parseParameter(SyntaxKind prevParamKind, boolean isParamNameOptio
             case AT_TOKEN:
                 annots = parseOptionalAnnotations();
                 break;
-            case PUBLIC_KEYWORD:
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1305]: sorry I didn't see this earlier.  Have to remove parsing of public keyword in
https://github.com/ballerina-platform/ballerina-lang/pull/25228/files#diff-51d1fcc2878d0934de8afd17c12009dcL1336-L1338

**Comment by lochana-chathura** [comment]
https://github.com/ballerina-platform/ballerina-lang/blob/603ef231ce54aed22951e96598144471fc9384d7/compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json#L886-L889

https://github.com/ballerina-platform/ballerina-lang/blob/603ef231ce54aed22951e96598144471fc9384d7/compiler/ballerina-treegen/src/main/resources/syntax_tree_descriptor.json#L920-L923

We can remove optional visibility qualifier for parameter nodes in the syntax tree right?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 4 above)
Comment (dulvinw) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1305]: Will fix this!

**Review by lochana-chathura — CHANGES_REQUESTED** [review]
Shall we do this small refactoring also :)

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -1333,16 +1333,11 @@ private STNode parseParamGivenAnnots(SyntaxKind prevParamKind, STNode annots, bo
...
-                break;
             case IDENTIFIER_TOKEN:
-                qualifier = STNodeFactory.createEmptyNode();
                 break;
             case AT_TOKEN: // Annotations can't reach here
             default:
                 if (isTypeStartingToken(nextToken.kind) && nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN) {
-                    qualifier = STNodeFactory.createEmptyNode();
                     break;
                 }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1342]: Since we don't parse any qualifiers here, this method simply does nothing. 
1. Can remove this entire method and direct call `parseParamGivenAnnotsAndQualifier()` beneath. 
2. rename: `parseParamGivenAnnotsAndQualifier` -> `parseParamGivenAnnots`

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 6]
```diff
@@ -2422,6 +2415,8 @@ protected ParserRuleContext getNextRule(ParserRuleContext currentCtx, int nextLo
             case DOUBLE_SLASH_DOUBLE_ASTERISK_LT_TOKEN:
             case SLASH_LT_TOKEN:
                 return ParserRuleContext.XML_NAME_PATTERN;
+            case PARAMETER_WITHOUT_ANNOTS:
+                return ParserRuleContext.TYPE_DESC_IN_PARAM;
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2419]: Once we remove above mentioned method. We don't need this recovery rule at all. 
Can remove all usages of `PARAMETER_WITHOUT_ANNOTS` in error handler and remove it from the ParserRuleContex.java as well.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 7]
```diff
@@ -1436,12 +1430,12 @@ private STNode parseEllipsis() {
      * @param paramName Name of the parameter
      * @return Parsed parameter node
      */
-    private STNode parseParameterRhs(SyntaxKind prevParamKind, STNode annots, STNode qualifier, STNode type,
+    private STNode parseParameterRhs(SyntaxKind prevParamKind, STNode annots, STNode type,
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1433]: Update the above doc params for this chage as well

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 8]
```diff
@@ -1329,54 +1329,21 @@ private STNode parseParameter(SyntaxKind prevParamKind, boolean isParamNameOptio
...
-            case AT_TOKEN: // Annotations can't reach here
-            default:
-                if (isTypeStartingToken(nextToken.kind) && nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN) {
-                    qualifier = STNodeFactory.createEmptyNode();
-                    break;
-                }
-
-                Solution solution = recover(nextToken, ParserRuleContext.PARAMETER_WITHOUT_ANNOTS, prevParamKind,
-                        annots, isParamNameOptional);
+    private STNode parseParamGivenAnnots(SyntaxKind prevParamKind, STNode annots,
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1332]: I think we no longer need this method. We can inline the content.

**Review by MaryamZi — APPROVED** [review]
Compiler FE and stdlib changes LGTM. We can merge when we're OK with the parser changes.

**Review by SupunS — APPROVED** [review]
